broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java

```diff
@@ -710,6 +663,57 @@ protected void handleFailure(final Exception failure) {
     onInstallFailure();
   }
 
+  private void addClosingStep(final String name, final AsyncClosable closable) {
+    closingSteps.add(new ClosingStep(name, closable));
+  }
+
+  private CompletableActorFuture<Void> closePartition() {
+    // this method may be called concurrently when the actor is closed
+    final var closingStepsInReverseOrder = new ArrayList<>(closingSteps);
+    Collections.reverse(closingStepsInReverseOrder);
+
+    final var closingPartitionFuture = new CompletableActorFuture<Void>();
+    stepByStepClosing(closingPartitionFuture, closingStepsInReverseOrder);
+
+    return closingPartitionFuture;
+  }
+
+  private void stepByStepClosing(
+      final CompletableActorFuture<Void> closingFuture, final List<ClosingStep> actorsToClose) {
+    if (actorsToClose.isEmpty()) {
+      closingFuture.complete(null);
+      return;
+    }
+
+    final ClosingStep closingStep = actorsToClose.remove(0);
+    LOG.debug("Closing Zeebe-Partition-{}: {}", partitionId, closingStep.getName());
+
+    final var closeFuture = closingStep.getClosable().closeAsync();
+    closeFuture.onComplete(
+        (v, t) -> {
+          if (t == null) {
+            LOG.debug(
+                "Closing Zeebe-Partition-{}: {} closed successfully",
+                partitionId,
+                closingStep.getName());
+
+            // remove the completed step from the list in case that the closing is interrupted
```

Comment:
Ideally we should prevent that; in case of multiple role changes we do this due to waiting of the last transition. I think it makes sense to intorduce a closing flag which is toggled and nothing can be executed anymore, like transition. I think actor scheduler also supports something like that. That some jobs are only executed on starting right?

Reply:
When `actor.close()` is called then every scheduled job is canceled. That mean, if `closePartition()` is called first and `actor.close()` is called later then this interrupts the `closeFuture.onComplete()` chain in `stepByStepClosing()`. So, we need to continue with the closing of the partition.

What exactly do you suggest?

Reply:
OK the comments before sounded like it is a problem when we are do the actor closing that the next transition interrupt it?

> When actor.close() is called then every scheduled job is canceled.
But partition closing if scheduled will not be interrupted?